pkg/controller/iosvariant/iosvariant_controller.go

```diff
@@ -42,7 +43,15 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 	}
 
 	// Watch for changes to primary resource IOSVariant
-	err = c.Watch(&source.Kind{Type: &pushv1alpha1.IOSVariant{}}, &handler.EnqueueRequestForObject{})
+	onlyEnqueueForValidNamespaces, err := nspredicate.NewFromEnvVar("APP_NAMESPACES")
```

Comment:
The implementation needs to be fone in the reconcile. The WATCH is just observers(triggers) in the cluster. It will not perform what is required here. Note that the solution applied in the MSS Oper is the same applied for other old projects maintained by us. See [here](https://github.com/aerogear/app-metrics-operator/pull/11/files#diff-2752e78fa6777e8cd072967fb0801b5fR99) how it was done in the PR for App Metrics.  

When we run `operator-sdk local up` or `unit-test` it is outside of the cluster and because of this, a specific error is faced when we call the framework features. Because of this, we are able to identify these scenarios and allow continue. See [here](https://github.com/aerogear/app-metrics-operator/pull/11/files#diff-3661dc32b6bde1b0939ad409364e48fcR1) its tests to ensure that works well as its implementation [here](https://github.com/aerogear/app-metrics-operator/pull/11/files#diff-fb81b0a96eed5b010cfb081b3aef4431R65)

Reply:
This seems like a much better place to check than in the reconcile method, and exactly the kind of thing that this is for. It works exactly as I expect (both inside and outside of the cluster, with the caveat that locally it seems to only be able to react to events in one namespace, but that was the case beforehand also).

> It will not perform what is required here.

Please elaborate. What do you see as being "required" here that will not be "performed"?

> it is outside of the cluster and because of this, a specific error is faced when we call the framework features.

Please elaborate. What "framework features" are being used here that won't work when running locally?

Reply:
👍 You added the validation in the primary one. Yep. It should work. You are correct.
However, is it adding in the logs that the namespace is invalid for the CR?

Reply:
Hmm, which command are you running to see that error? The existing tests appear to work fine for me on this branch:

```
make test/unit 
Running tests:
CGO_ENABLED=1 go test -v -race -cover github.com/aerogear/unifiedpush-operator/pkg/controller/androidvariant
=== RUN   TestReconcileAndroidVariant_Reconcile
--- PASS: TestReconcileAndroidVariant_Reconcile (0.01s)
PASS
coverage: 16.1% of statements
ok  	github.com/aerogear/unifiedpush-operator/pkg/controller/androidvariant	1.157s	coverage: 16.1% of statements
```

Reply:
It is working fine because in this test we are not checking the objects managed and created by it. It was added just as an example.

Reply:
@grdryn let's merge the following PR then we can check if it will work or not properly. 

https://github.com/aerogear/unifiedpush-operator/pull/47